File: generate.py
import os
import subprocess
import sys
import re

svg_folders = ["line", "solid", "thinline", "monochrome"]

def run_command(command):
    """Utility function to run a shell command."""
    result = subprocess.run(command, shell=True, check=True)
    if result.returncode != 0:
        print(f"Command failed: {command}", file=sys.stderr)
        sys.exit(result.returncode)

def process_svgs():
    """Process SVG folders using svgr and custom templates."""
    for folder in svg_folders:
        print(f"Processing '{folder}' style icons...")
        command = (
            f"pnpm svgr "
            f"--index-template=templates/indexTemplate.cjs "
            f"--template=templates/iconTemplate.cjs "
            f"--svg-props=\"height={{size}},width={{size}},fill={{color}}\" "
            f"--ref "
            f"--typescript "
            f"--filename-case kebab "
            f"--out-dir generated/{folder} "
            f"-- node_modules/@iconscout/unicons/svg/{folder}"
        )
        run_command(command)
        print(f"Generated folder for '{folder}' style icons")

def replace_classnames_in_monochrome():
    """Replace classNames with opacity props in monochrome style icons."""
    print("Replacing classNames in 'monochrome' style icons...")

    monochrome_dir = "generated/monochrome"
    for filename in os.listdir(monochrome_dir):
        if filename.endswith(".tsx"):
            filepath = os.path.join(monochrome_dir, filename)

            with open(filepath, "r") as file:
                content = file.read()

            # Delete existing opacity={...} props
            content = re.sub(r'opacity={.*}', '', content)

            # Replace classNames with corresponding opacity props
            content = re.sub(r'className="uim-primary"', 'opacity={1}', content)
            content = re.sub(r'className="uim-secondary"', 'opacity={0.7}', content)
            content = re.sub(r'className="uim-tertiary"', 'opacity={0.5}', content)
            content = re.sub(r'className="uim-quaternary"', 'opacity={0.25}', content)

            with open(filepath, "w") as file:
                file.write(content)

    print("Finished replacing classNames in 'monochrome' style icons")

def generate_unicon_props():
    """Generate UniconProps.ts file."""
    unicon_props_content = """
import type { SVGProps } from 'react';

export interface UniconProps extends SVGProps<SVGSVGElement> {
    /**
     * The size of the icon.
     * Passed down to the SVG's height and width attributes.
     * @default 24
     */
    size?: number | string;

    /**
     * The color of the icon.
     * @default 'currentColor'
     */
    color?: string;
}
"""
    unicon_props_path = "generated/UniconProps.ts"
    with open(unicon_props_path, "w") as file:
        file.write(unicon_props_content.strip())
    print("Generated UniconProps.ts")

def main():
    os.makedirs("generated", exist_ok=True)
    process_svgs()
    replace_classnames_in_monochrome()
    generate_unicon_props()

if __name__ == "__main__":
    main()

File: generate.js
import fs from "fs";
import path from "path";
import { execSync } from "child_process";

const svgFolders = ["line", "solid", "thinline", "monochrome"];

function runCommand(command) {
  try {
    execSync(command, { stdio: 'inherit' });
  } catch (error) {
    console.error(`Command failed: ${command}`);
    process.exit(1);
  }
}

function processSvgs() {
  svgFolders.forEach(folder => {
    console.log(`Processing '${folder}' style icons...`);
    const command = `
      pnpm svgr \
      --index-template=templates/indexTemplate.cjs \
      --template=templates/iconTemplate.cjs \
      --svg-props="height={size},width={size},fill={color}" \
      --ref \
      --typescript \
      --filename-case kebab \
      --out-dir generated/${folder} \
      -- node_modules/@iconscout/unicons/svg/${folder}
    `;
    runCommand(command);
    console.log(`Generated folder for '${folder}' style icons`);
  });
}

function replaceClassnamesInMonochrome() {
  console.log("Replacing classNames in 'monochrome' style icons...");

  const monochromeDir = path.join("generated", "monochrome");

  // Check if the directory exists before trying to read it
  if (!fs.existsSync(monochromeDir)) {
    console.warn(`Directory '${monochromeDir}' does not exist. Skipping className replacements.`);
    return;
  }

  const files = fs.readdirSync(monochromeDir);

  files.forEach(filename => {
    if (filename.endsWith(".tsx")) {
      const filepath = path.join(monochromeDir, filename);
      let content = fs.readFileSync(filepath, 'utf-8');

      // Delete existing opacity={...} props
      content = content.replace(/opacity={.*}/g, '');

      // Replace classNames with corresponding opacity props
      content = content.replace(/className="uim-primary"/g, 'opacity={1}');
      content = content.replace(/className="uim-secondary"/g, 'opacity={0.7}');
      content = content.replace(/className="uim-tertiary"/g, 'opacity={0.5}');
      content = content.replace(/className="uim-quaternary"/g, 'opacity={0.25}');

      fs.writeFileSync(filepath, content, 'utf-8');
    }
  });

  console.log("Finished replacing classNames in 'monochrome' style icons");
}

function generateUniconProps() {
  const uniconPropsContent = `
import type { SVGProps } from 'react';

export interface UniconProps extends SVGProps<SVGSVGElement> {
    /**
     * The size of the icon.
     * Passed down to the SVG's height and width attributes.
     * @default 24
     */
    size?: number | string;

    /**
     * The color of the icon.
     * @default 'currentColor'
     */
    color?: string;
}
  `;

  const uniconPropsPath = path.join("generated", "UniconProps.ts");
  fs.writeFileSync(uniconPropsPath, uniconPropsContent.trim(), 'utf-8');
  console.log("Generated UniconProps.ts");
}

function main() {
  if (!fs.existsSync("generated")) {
    fs.mkdirSync("generated");
  }
  processSvgs();
  replaceClassnamesInMonochrome();
  generateUniconProps();
}

main();
